util/filesort/filesort.go

```diff
@@ -0,0 +1,392 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filesort
+
+import (
+	"container/heap"
+	"encoding/binary"
+	"io"
+	"os"
+	"path"
+	"sort"
+	"strconv"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type comparableRow struct {
+	key    []types.Datum
+	val    []types.Datum
+	handle int64
+}
+
+type item struct {
+	index int // source file index
+	value *comparableRow
+}
+
+// Min-heap of comparableRows
+type rowHeap struct {
+	sc     *variable.StatementContext
+	ims    []*item
+	byDesc []bool
+}
+
+func lessThan(sc *variable.StatementContext, i []types.Datum, j []types.Datum, byDesc []bool) bool {
+	for k := range byDesc {
+		v1 := i[k]
+		v2 := j[k]
+
+		ret, err := v1.CompareDatum(sc, v2)
+		if err != nil {
+			panic(err)
```

Comment:
We should not panic here.
We can add an error field in `rowHeap` and make this function a method of `rowHeap`, then if there is an error occurred, set error and return false. 
After sort, we can check and return the error.